compiler/rustc_middle/src/mir/mono.rs

```diff
@@ -103,17 +102,13 @@ impl<'tcx> MonoItem<'tcx> {
                 // inlined function. If we're inlining into all CGUs then we'll
                 // be creating a local copy per CGU.
                 if generate_cgu_internal_copies {
-                    return InstantiationMode::LocalCopy;
-                }
-
-                // Finally, if this is `#[inline(always)]` we're sure to respect
-                // that with an inline copy per CGU, but otherwise we'll be
-                // creating one copy of this `#[inline]` function which may
-                // conflict with upstream crates as it could be an exported
-                // symbol.
-                match tcx.codegen_fn_attrs(instance.def_id()).inline {
-                    InlineAttr::Always => InstantiationMode::LocalCopy,
-                    _ => InstantiationMode::GloballyShared { may_conflict: true },
+                    InstantiationMode::LocalCopy
+                } else {
+                    // Finally, if we've reached this point, then we know that OptLevel = No so we
```

Comment:
Can we put an assertion for it here?

Reply:
Done!

Reply:
It fails with ...

```
warning: function is never used: `const_fn`
  --> /checkout/src/test/codegen/link-dead-code.rs:11:10
   |
11 | const fn const_fn() -> i32 { 1 }
   |          ^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: function is never used: `inline_fn`
  --> /checkout/src/test/codegen/link-dead-code.rs:17:4
   |
17 | fn inline_fn() -> i32 { 2 }
   |    ^^^^^^^^^

warning: function is never used: `private_fn`
  --> /checkout/src/test/codegen/link-dead-code.rs:22:4
   |
22 | fn private_fn() -> i32 { 3 }
   |    ^^^^^^^^^^

thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `Default`,
 right: `No`', compiler/rustc_middle/src/mir/mono.rs:111:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.49.0-nightly (6dd067e3b 2020-10-23) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z unstable-options -C prefer-dynamic -C rpath -C debuginfo=0 -C link-dead-code
```

Removing that commit from this PR but unsure why is `Default` for that test case and not `No`

Reply:
So it's not always `No`, it could be something else if `inline_in_all_cgus` is `false` or there's the `!tcx.sess.link_dead_code()` like in the tests that is now failing.

Reply:
I've adjusted the comment.